### PR TITLE
mkl: 2020.3.304 -> 2021.1.1.52

### PR DESCRIPTION
**Pull request by danieldk**
###### Motivation for this change

Intel MKL is succeeded by Intel oneMKL. This also results in several
changes:

- The versioning scheme has been changed, it now consists of a version
  such as "2021.1.1" and a build such as "52".
- The tarball of RPMs has been replaced by a relatively complex
  installer, so instead we use the RPMs provided by the upstream
  Yum/DNF repository.

To better detect breakage, the passthru tests also check static builds
and with parallel (iomp) execution.

Tested: PyTorch build.

Note: some of the libraries have unresolved dynamic library references. We can't really solve them, since it would require `patchelf`-ing the libraries (which constitutes modification under the license?). However, this was also already true in the previous version. Consequently, it's up to downstream users to link these libraries.

Marked WIP to do more testing.

cc: @bhipple 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
> Note: some of the libraries have unresolved dynamic library references. We can't really solve them, since it would require patchelf-ing the libraries (which constitutes modification under the license?). However, this was also already true in the previous version. Consequently, it's up to downstream users to link these libraries.

If they are not patched then they are not usable on a nixos system. If we don't hydra cache the results the end user needs to do it anyway.

**Comment by danieldk** [comment]
> If they are not patched then they are not usable on a nixos system.

This is incorrect. It's just that the burden to link those libraries moves to downstream derivations/applications. However, this has always been like this with the MKL derivation, we can't patch it because because the license does not allow it.

However, in practice it is not a problem, because for the default threading options (sequential and OpenMP) do not have external dependencies outside libc. For that reason, we can build numpy or PyTorch against MKL just fine.

The problematic runtimes are Thread Building Blocks (requires libhwloc and libstdc++) and SYCL (requires libsycl). However, I think it still useful to provide these as well, because they can be used by downstream users by linking the requisite libraries.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 117354` at 6556beb7 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built successfully:</summary>
  <ul>
    <li>clmagma</li>
    <li>mkl</li>
    <li>python38Packages.mkl-service</li>
    <li>python39Packages.mkl-service</li>
  </ul>
</details>

File: pkgs/development/libraries/science/math/mkl/default.nix [hunk 1]
```diff
@@ -17,40 +17,57 @@
...
+
+  oneapi-openmp = fetchurl {
+    url = "https://yum.repos.intel.com/oneapi/intel-oneapi-openmp-${mklVersion}-${mklVersion}-${openmpRel}.x86_64.rpm";
+    hash = "sha256-yP2c4aQAFNRffjLoIZgWXLcNXbiez8smsgu2wXitefU=";
+  };
+
+  oneapi-tbb = fetchurl {
+    url = "https://yum.repos.intel.com/oneapi/intel-oneapi-tbb-${mklVersion}-${mklVersion}-${tbbRel}.x86_64.rpm";
+    hash = "sha256-K1BvhGoGVU2Zwy5vg2ZvJWBrSdh5uQwo0znt5039X0A=";
+  };
```
Comment (bhipple) [inline pkgs/development/libraries/science/math/mkl/default.nix:64]: How does this interact with https://github.com/NixOS/nixpkgs/blob/d600f006643e074c2ef1d72e462e218b647a096c/pkgs/development/libraries/tbb/default.nix?

Comment (danieldk) [reply pkgs/development/libraries/science/math/mkl/default.nix:64]: That's a good question. We should verify if it is ABI compatible with the version that MKL is compiled with. The other issue is that we'd have to set `RUNPATH` to actually resolve that dependency in the library, which requires modification.

**Review by bhipple — APPROVED** [review]
I'd prefer if we could resolve the `tbb` duplication, but that seems OK to do in a follow-up.  Not being able to run `patchelf` on the binary RPaths due to licensing restrictions makes this tricky.

**Pull request merged**